hazelcast/include/hazelcast/client/serialization/pimpl/compact.h

```diff
@@ -0,0 +1,426 @@
+    static exception::hazelcast_serialization unexpected_null_value(
+      const std::string& field_name,
+      const std::string& method_suffix);
+    size_t read_fixed_size_position(
+      const pimpl::field_descriptor& field_descriptor) const;
+    size_t read_var_size_position(
+      const pimpl::field_descriptor& field_descriptor) const;
+    pimpl::compact_stream_serializer& compact_stream_serializer;
+    serialization::object_data_input& object_data_input;
+    const pimpl::schema& schema;
+    uint32_t data_start_position;
+    uint32_t variable_offsets_position;
+    std::function<int(serialization::object_data_input&, uint32_t, uint32_t)>
+      get_offset;
+};
+
+/**
+ *  Provides means of writing compact serialized fields to the binary data.
+ */
+class HAZELCAST_API compact_writer
+{
+public:
+    /**
+     * Writes a 32-bit two's complement signed integer.
+     *
+     * @param fieldName name of the field.
+     * @param value     to be written.
+     */
+    void write_int32(const std::string& field_name, int32_t value);
+
+    /**
+     * Writes an UTF-8 encoded string.
+     *
+     * @param fieldName name of the field.
+     * @param value     to be written.
+     */
+    void write_string(const std::string& field_name,
+                      const boost::optional<std::string>& value);
+
+private:
+    friend compact_writer pimpl::create_compact_writer(
+      pimpl::default_compact_writer* default_compact_writer);
+    friend compact_writer pimpl::create_compact_writer(
+      pimpl::schema_writer* schema_writer);
+    explicit compact_writer(
+      pimpl::default_compact_writer* default_compact_writer);
+    explicit compact_writer(pimpl::schema_writer* schema_writer);
+    pimpl::default_compact_writer* default_compact_writer;
+    pimpl::schema_writer* schema_writer;
+};
+
+namespace pimpl {
+
+namespace offset_reader {
+/**
+ * Offset of the null fields.
+ */
+constexpr int32_t NULL_OFFSET = -1;
+/**
+ * Range of the offsets that can be represented by a single byte
+ * and can be read with offset_reader::BYTE_OFFSET_READER.
+ */
+constexpr uint32_t BYTE_OFFSET_READER_RANGE = INT8_MAX - INT8_MIN;
+
+/**
+ * Range of the offsets that can be represented by two bytes
+ * and can be read with offset_reader#SHORT_OFFSET_READER}.
+ */
+constexpr uint32_t SHORT_OFFSET_READER_RANGE = INT16_MAX - INT16_MIN;
+
+/**
+ * Returns the offset of the variable-size field at the given index.
+ *
+ * @tparam OFFSET_TYPE can be int32_t, int16_t or int8_t
+ * @param in Input to read the offset from.
+ * @param variableOffsetsPos Start of the variable-size field offsets
+ *                           section of the input.
+ * @param index Index of the field.
+ * @return The offset.
+ */
+template<typename OFFSET_TYPE>
+int32_t
+get_offset(serialization::object_data_input& in,
+           uint32_t variable_offsets_pos,
+           uint32_t index);
+
+} // namespace offset_reader
+
+class HAZELCAST_API default_compact_writer
```

Comment:
Why don't we have default_compact_reader if we have default_compact_writer?  I think we can put implementation and the interface to the same class in cpp. (like compact_reader above)